src/signed.rs

```diff
@@ -1394,6 +1396,58 @@ where
     }
 }
 
+macro_rules! bytes_operation_impl {
+    ($base_data_type:ty, $bits:expr, [$($indices:expr),+]) => {
+        impl Int<$base_data_type, $bits>
+        {
+            pub const fn to_le_bytes(&self) -> [u8; $bits >> 3] {
+                let v = self.value();
+
+                [ $( (v >> ($indices << 3)) as u8, )+ ]
+            }
+
+            pub const fn from_le_bytes(from: [u8; $bits >> 3]) -> Self {
+                let value = { 0 $( | (from[$indices] as $base_data_type) << ($indices << 3))+ };
+                Self { value: (value << Self::UNUSED_BITS) >> Self::UNUSED_BITS }
+            }
+
+            pub const fn to_be_bytes(&self) -> [u8; $bits >> 3] {
+                 let v = self.value();
+
+                [ $( (v >> ($bits - 8 - ($indices << 3))) as u8, )+ ]
+            }
+
+            pub const fn from_be_bytes(from: [u8; $bits >> 3]) -> Self {
+                let value = { 0 $( | (from[$indices] as $base_data_type) << ($bits - 8 - ($indices << 3)))+ };
+                Self { value: (value << Self::UNUSED_BITS) >> Self::UNUSED_BITS }
```

Comment:
According to the Compiler Explorer doing this manually produces suboptimal code if the target architecture is little endian, even with all optimisations enabled (tested with `UInt::<u32, 24>` but it should apply here as well): https://godbolt.org/z/zY1M91aGz

Any chance we could do something like this instead? It should be ever so slightly faster if there's a need to endian-swap, plus it saves a shift with signed numbers:
```suggestion
                let mut bytes = [0_u8; size_of::<$base_data_type>()];
                // Adding the amount of unused bytes to the index makes the sign bit show up at the underlying type's MSB, so a right shift can sign-extend the value into place. This only works because `Self::UNUSED_BITS` is always a multiple of 8.
                $(bytes[(Self::UNUSED_BITS / 8) + $indices] = from[$indices];)+
                Self { value: <$base_data_type>::from_be_bytes(bytes) >> Self::UNUSED_BITS }
```
This should for `UInt` as well if the offset into `bytes` is removed.

Reply:
Great idea. Seems to work pretty efficiently up to u64 - at u128 it starts introducing shifts, but a future compiler might fix that

Reply:
I really like your new approach, I had some difficulty comprehending the previous solution but with your changes its much more intuitive :)

Too bad it doesn't work on the old nightly for `const_convert_and_const_trait_impl`, though supporting that doesn't really seem worth forever restricting this crate to Rust features from 2023. Out of curiosity, what doesn't it support?

Reply:
Mutable references in const functions. To support the old compiler, the copy bytes function would need to be inlined